src/openslide-vendor-kfbio.c

```diff
@@ -0,0 +1,751 @@
+  }
+  int64_t label_data_in_file = label_info_in_file + 52;
+  if (!_openslide_jpeg_add_associated_image(osr, "label", filename, label_data_in_file, err))
+  {
+    g_prefix_error(err, "Couldn't read associated image: %s", "label");
+    return false;
+  }
+
+  if (!_openslide_fseek(f, preview_info_in_file, SEEK_SET, err))
+  {
+    g_prefix_error(err, "Couldn't seek within header: ");
+    return false;
+  }
+  if (!_openslide_fseek(f, 8, SEEK_CUR, err))
+  {
+    g_prefix_error(err, "Couldn't seek within header: ");
+    return false;
+  }
+  int32_t preview_h = read_le_int32_from_file(f);
+  g_assert(preview_h > 0);
+  int32_t preview_w = read_le_int32_from_file(f);
+  g_assert(preview_w > 0);
+  if (!_openslide_fseek(f, 4, SEEK_CUR, err))
+  {
+    g_prefix_error(err, "Couldn't seek within header: ");
+    return false;
+  }
+  int32_t preview_length = read_le_int32_from_file(f);
+  g_assert(preview_length > 0);
+  if (!_openslide_fseek(f, 28, SEEK_CUR, err))
+  {
+    g_prefix_error(err, "Couldn't seek within header: ");
+    return false;
+  }
+  int64_t preview_data_in_file = preview_info_in_file + 52;
+  if (!_openslide_jpeg_add_associated_image(osr, "preview", filename, preview_data_in_file, err))
+  {
+    g_prefix_error(err, "Couldn't read associated image: %s", "preview");
+    return false;
+  }
+
+  // set up level dimensions and such
+  g_autoptr(GPtrArray) level_array =
+      g_ptr_array_new_with_free_func((GDestroyNotify)destroy_level);
+  int64_t downsample = 1;
+  for (int i = 0; i < zoom_levels; i++)
+  {
+    // ensure downsample is > 0 and a power of 2
+    if (downsample <= 0 || (downsample & (downsample - 1)))
+    {
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "Invalid downsample %" PRId64, downsample);
+      return false;
+    }
+
+    if (i == 0)
+      downsample = 1;
+    else
+      downsample *= 2;
+
+    struct level *l = g_new0(struct level, 1);
+    g_ptr_array_add(level_array, l);
+
+    l->base.downsample = downsample;
+    l->base.tile_w = (double)tile_size;
+    l->base.tile_h = (double)tile_size;
+
+    l->base.w = base_w / l->base.downsample;
+    if (l->base.w == 0)
+      l->base.w = 1;
+    l->base.h = base_h / l->base.downsample;
+    if (l->base.h == 0)
+      l->base.h = 1;
+
+    l->grid = _openslide_grid_create_tilemap(osr,
+                                             tile_size,
+                                             tile_size,
+                                             read_tile, tile_free);
+  }
+
+  // load the position map and build up the tiles
+  if (!process_tiles_info_from_header(f,
+                                      tiles_info_in_file,
+                                      zoom_levels,
+                                      tile_count,
+                                      (struct level **)level_array->pdata,
+                                      err))
+  {
+    return false;
+  }
+
+  for (int i = 0; i < zoom_levels; i++)
```

Comment:
I don't fully understand what is this block doing or it has some mistake.
You are checking `base.tile_w` and `base.tile_h` values for each level. But all those values were set only once (line 686, 687) to `tile_size`, so you can do the check right after you parsed `tile_size` from the file.

Reply:
You are right, will remove this block as it is not really necessary.